codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -239,48 +239,49 @@ private void processPayloadState(ChannelHandlerContext ctx, ByteBuf in, Http2Fra
             return;
         }
 
-        // Get a view of the buffer for the size of the payload.
-        ByteBuf payload = in.readSlice(payloadLength);
+        // Only process up to payloadLength bytes, this is the excess
+        int remaining = in.readableBytes() - payloadLength;
 
         // We have consumed the data, next time we read we will be expecting to read a frame header.
         readingHeaders = true;
 
         // Read the payload and fire the frame event to the listener.
         switch (frameType) {
             case DATA:
-                readDataFrame(ctx, payload, listener);
+                readDataFrame(ctx, in, remaining, listener);
                 break;
             case HEADERS:
-                readHeadersFrame(ctx, payload, listener);
+                readHeadersFrame(ctx, in, remaining, listener);
                 break;
             case PRIORITY:
-                readPriorityFrame(ctx, payload, listener);
+                readPriorityFrame(ctx, in, listener);
                 break;
             case RST_STREAM:
-                readRstStreamFrame(ctx, payload, listener);
+                readRstStreamFrame(ctx, in, listener);
                 break;
             case SETTINGS:
-                readSettingsFrame(ctx, payload, listener);
+                readSettingsFrame(ctx, in, listener);
                 break;
             case PUSH_PROMISE:
-                readPushPromiseFrame(ctx, payload, listener);
+                readPushPromiseFrame(ctx, in, remaining, listener);
                 break;
             case PING:
-                readPingFrame(ctx, payload.readLong(), listener);
+                readPingFrame(ctx, in.readLong(), listener);
                 break;
             case GO_AWAY:
-                readGoAwayFrame(ctx, payload, listener);
+                readGoAwayFrame(ctx, in, remaining, listener);
                 break;
             case WINDOW_UPDATE:
-                readWindowUpdateFrame(ctx, payload, listener);
+                readWindowUpdateFrame(ctx, in, listener);
                 break;
             case CONTINUATION:
-                readContinuationFrame(payload, listener);
+                readContinuationFrame(in, remaining, listener);
                 break;
             default:
-                readUnknownFrame(ctx, payload, listener);
+                readUnknownFrame(ctx, in, remaining, listener);
                 break;
         }
+        in.readerIndex(in.writerIndex() - remaining);
```

Comment:
nit: I think it may be easier to read to use `in.skipBytes(payloadLength);`

Reply:
@normanmaurer I don't think this will work since the reader index is now moved inside the above methods (since `in` now isn't sliced)

Reply:
ah sorry my bad...